R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
 
   out <- nq / n * out
 }
+
+#' @title (Internal) Design-based variance estimates with HC0 adjustment
+#' @param x a fitted \code{teeMod} model
+#' @details The design-based variance estimates can be calculated for 
+#' \code{teeMod} models satisfying the following requirements:
+#' - The model uses \code{rct_design} as \code{Design}
+#' - The model only estimates a main treatment effect
+#' - Inverse probability weighting is incorporated
+#' 
+#' @keywords internal
+#' @rdname var_estimators
+.vcov_DB0 <- function(x, ...) {
+  if (!inherits(x, "teeMod")) {
+    stop("x must be a teeMod model")
+  }
+  
+  if (!x@lmitt_fitted){
+    # x@lmitt_fitted is false if someone created x using as.lmitt
+    stop("x must have been fitted using lmitt.formula")
+  }
+  
+  if (inherits(os <- x$model$`(offset)`, "SandwichLayer"))
+    if (!all(os@keys$in_Q)){
+      stop(paste("Design-based standard errors cannot be computed for teeMod",
+                 "models with external sample for covariance adjustment"))
+    }
+  
+  if (x@Design@type != "RCT"){
+    stop("x must be an RCT design")
+  }
+  
+  if (length(x@moderator) > 0){
+    stop(paste("Design-based standard errors cannot be computed for teeMod",
+               "models with moderators"))
+  }
+  
+  args <- list(...)
+  if ("type" %in% names(args)) {
+    stop(paste("Cannot override the `type` argument for meat",
+               "matrix computations"))
+  }
+  args$x <- x
+  args$db <- TRUE
+  n <- length(args$cluster)
+  
+  if (x@absorbed_intercepts) {
+    a22inv <- sandwich::bread(x)
+    meat <- do.call(sandwich::meatCL, args)
+    
+    vmat <- (1 / n) * a22inv %*% meat %*% a22inv
+    name <- grep("z", colnames(x$model))
+    vmat <- as.matrix(vmat[name, name])
+  }
+  else {
+    # if ate() is not called for weights, throw a warning
+    if (sum(grepl("ate", x@lmitt_call$weights)) == 0){
```

Comment:
It seems that `x@lmitt_call$weights[[1]]` is `"*"` instead of `"ate"` when `weights` is specified as something like `ate(des) * data$weight` in `lmitt()`. Changed this logic to `if (!(inherits(x@lmitt_call$weights, "call") & 
          sum(grepl("ate", x@lmitt_call$weights)) > 0))`